homeassistant/components/ecovacs/__init__.py

```diff
@@ -1,91 +1,120 @@
+            LOGGER.info(
+                "Shutting down connection to Ecovacs device %s",
+                device.vacuum[ECOVACS_ATTR_DEVICE_ID],
+            )
+            device.disconnect()
 
-    ecovacs_api = EcoVacsAPI(
-        ECOVACS_API_DEVICEID,
-        config[DOMAIN].get(CONF_USERNAME),
-        EcoVacsAPI.md5(config[DOMAIN].get(CONF_PASSWORD)),
-        config[DOMAIN].get(CONF_COUNTRY),
-        config[DOMAIN].get(CONF_CONTINENT),
+    ecovacs_api = await hass.async_add_executor_job(
+        partial(
+            EcoVacsAPI,
+            entry.data[CONF_DEVICE_ID],
+            entry.data[CONF_USERNAME],
+            EcoVacsAPI.md5(entry.data[CONF_PASSWORD]),
+            entry.data[CONF_COUNTRY],
+            entry.data[CONF_CONTINENT],
+        )
     )
 
-    devices = ecovacs_api.devices()
-    _LOGGER.debug("Ecobot devices: %s", devices)
+    vacuums = []
+    devices = await hass.async_add_executor_job(ecovacs_api.devices)
 
     for device in devices:
-        _LOGGER.info(
+        LOGGER.info(
             "Discovered Ecovacs device on account: %s with nickname %s",
-            device["did"],
-            device["nick"],
+            device[ECOVACS_ATTR_DEVICE_ID],
+            device[ECOVACS_ATTR_NAME],
         )
         vacbot = VacBot(
             ecovacs_api.uid,
             ecovacs_api.REALM,
             ecovacs_api.resource,
             ecovacs_api.user_access_token,
             device,
-            config[DOMAIN].get(CONF_CONTINENT).lower(),
+            entry.data[CONF_CONTINENT],
             monitor=True,
         )
-        hass.data[ECOVACS_DEVICES].append(vacbot)
+        vacuums.append(vacbot)
 
-    def stop(event: object) -> None:
-        """Shut down open connections to Ecovacs XMPP server."""
-        for device in hass.data[ECOVACS_DEVICES]:
-            _LOGGER.info(
-                "Shutting down connection to Ecovacs device %s", device.vacuum["did"]
-            )
-            device.disconnect()
+    remove_stop_listener = hass.bus.async_listen_once(
+        EVENT_HOMEASSISTANT_STOP, stop_ecovacs
+    )
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = {
+        DEVICES: vacuums,
+        DATA_REMOVE_LISTENER: remove_stop_listener,
+    }
+
+    for component in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+
+    return True
+
+
+async def async_unload_entry(hass, entry):
+    """Unloading the Ecovacs platforms."""
+
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
+
+    if not unload_ok:
+        return False
 
-    # Listen for HA stop to disconnect.
-    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, stop)
+    hass.data[DOMAIN][entry.entry_id][DATA_REMOVE_LISTENER]()
```

Comment:
Not sure if I have implemented this listener correctly, since the attached method does not log disconnecting from its devices. However, when stopping HA the attached listener method logs successfully.

Reply:
The listener is listening to the EVENT_HOMEASSISTANT_STOP, you cancel the listener by calling it. So you might be lucky timing wise, you could set up a second listener that you can call from here instead